storage/innobase/btr/btr0pcur.cc

```diff
@@ -214,102 +212,98 @@ btr_pcur_copy_stored_position(
                                              mtr_t *mtr)
 {
-  ut_ad(block->page.buf_fix_count());
-  ut_ad(block->page.in_file());
-  ut_ad(block->page.frame);
-
   static_assert(BTR_SEARCH_PREV & BTR_SEARCH_LEAF, "");
   static_assert(BTR_MODIFY_PREV & BTR_MODIFY_LEAF, "");
   static_assert((BTR_SEARCH_PREV ^ BTR_MODIFY_PREV) ==
                 (RW_S_LATCH ^ RW_X_LATCH), "");
 
   const rw_lock_type_t mode=
     rw_lock_type_t(*latch_mode & (RW_X_LATCH | RW_S_LATCH));
+  buf_block_t *block= pcur->btr_cur.page_cur.block;
+  const page_id_t id{pcur->old_page_id};
+
+  if (!buf_page_optimistic_get(mode, block, id, pcur->modify_clock, mtr))
+    return false;
 
   switch (*latch_mode) {
   default:
     ut_ad(*latch_mode == BTR_SEARCH_LEAF || *latch_mode == BTR_MODIFY_LEAF);
-    return buf_page_optimistic_get(mode, block, pcur->modify_clock, mtr);
+    return true;
   case BTR_SEARCH_PREV:
   case BTR_MODIFY_PREV:
-    page_id_t id{0};
-    uint32_t left_page_no;
-    ulint zip_size;
-    buf_block_t *left_block= nullptr;
+    const auto savepoint= mtr->get_savepoint();
+    const page_t *page= block->page.frame;
+    const uint32_t left_page_no= btr_page_get_prev(page);
+    if (left_page_no == FIL_NULL)
+      return true;
+    const page_id_t left_page_id{id.space(), left_page_no};
+
+    /* We are holding a latch on the current page.
+
+    We will start by buffer-fixing the left sibling. Waiting for a
+    latch on it while holding a latch on the current page could lead
+    to a deadlock, because another thread could hold that latch and
+    wait for a right sibling page latch (the current page).
+
+    If there is a conflict, we will temporarily release our latch on
+    the current block while waiting for a latch on the left sibling.
+    The buffer-fixes on both blocks will prevent eviction. */
+
+    if (buf_block_t *left=
+        buf_page_get_gen(left_page_id, block->zip_size(), RW_NO_LATCH,
+                         nullptr, BUF_GET_POSSIBLY_FREED, mtr))
     {
-      transactional_shared_lock_guard<block_lock> g{block->page.lock};
-      if (block->modify_clock != pcur->modify_clock)
-        return false;
-      id= block->page.id();
-      zip_size= block->zip_size();
-      left_page_no= btr_page_get_prev(block->page.frame);
-    }
-
-    if (left_page_no != FIL_NULL)
-    {
-      left_block=
-        buf_page_get_gen(page_id_t(id.space(), left_page_no), zip_size,
-                         mode, nullptr, BUF_GET_POSSIBLY_FREED, mtr);
+      static_assert(MTR_MEMO_S_LOCK == BTR_SEARCH_LEAF << 5, "");
+      static_assert(MTR_MEMO_X_LOCK == BTR_MODIFY_LEAF << 5, "");
 
-      if (!left_block);
-      else if (btr_page_get_next(left_block->page.frame) != id.page_no())
+      if (mode == RW_S_LATCH
+          ? left->page.lock.s_lock_try() : left->page.lock.x_lock_try())
       {
-release_left_block:
-        mtr->release_last_page();
-        return false;
+        mtr->lock_register(savepoint + 1, mtr_memo_type_t(mode << 5));
```

Comment:
b56d2eb41733a2ae7daa3547928b8001838bccb3 addresses this (in another branch).

Reply:
There is still something wrong with `btr_pcur_move_backward_from_page()`. I hope to get an `rr replay` trace for that soon.

Reply:
9b169a63666f8a850c2d10e224ac9f6a6e39ae1f hopefully addresses this by invoking `btr_latch_prev()` and by improving that function so that the preceding block is always `mtr->at_savepoint(1)`.